smack-experimental/src/main/java/org/jivesoftware/smackx/stanza_content_encryption/provider/ContentElementProvider.java

```diff
@@ -0,0 +1,145 @@
+ * limitations under the License.
+ */
+package org.jivesoftware.smackx.stanza_content_encryption.provider;
+
+import java.io.IOException;
+import java.util.Date;
+
+import org.jivesoftware.smack.packet.XmlEnvironment;
+import org.jivesoftware.smack.parsing.SmackParsingException;
+import org.jivesoftware.smack.parsing.StandardExtensionElementProvider;
+import org.jivesoftware.smack.provider.ExtensionElementProvider;
+import org.jivesoftware.smack.provider.ProviderManager;
+import org.jivesoftware.smack.util.ParserUtils;
+import org.jivesoftware.smack.xml.XmlPullParser;
+import org.jivesoftware.smack.xml.XmlPullParserException;
+import org.jivesoftware.smackx.stanza_content_encryption.element.ContentElement;
+import org.jivesoftware.smackx.stanza_content_encryption.element.FromAffixElement;
+import org.jivesoftware.smackx.stanza_content_encryption.element.PayloadElement;
+import org.jivesoftware.smackx.stanza_content_encryption.element.RandomPaddingAffixElement;
+import org.jivesoftware.smackx.stanza_content_encryption.element.TimestampAffixElement;
+import org.jivesoftware.smackx.stanza_content_encryption.element.ToAffixElement;
+
+import org.jxmpp.jid.impl.JidCreate;
+import org.jxmpp.stringprep.XmppStringprepException;
+
+public class ContentElementProvider extends ExtensionElementProvider<ContentElement> {
+
+    @Override
+    public ContentElement parse(XmlPullParser parser, int initialDepth, XmlEnvironment xmlEnvironment)
+            throws XmlPullParserException, IOException, SmackParsingException {
+        ContentElement.Builder builder = ContentElement.builder();
+
+        boolean inContentElement = true;
+        while (inContentElement) {
+
+            XmlPullParser.Event tag = parser.next();
+            if (tag == XmlPullParser.Event.START_ELEMENT) {
+                String name = parser.getName();
+                switch (name) {
+                    case ToAffixElement.ELEMENT:
+                        parseToAffix(parser, builder);
+                        break;
+
+                    case FromAffixElement.ELEMENT:
+                        parseFromAffix(parser, builder);
+                        break;
+
+                    case TimestampAffixElement.ELEMENT:
+                        parseTimestampAffix(parser, builder);
+                        break;
+
+                    case RandomPaddingAffixElement.ELEMENT:
+                        parseRPadAffix(parser, builder);
+                        break;
+
+                    case PayloadElement.ELEMENT:
+                        parsePayload(parser, builder);
+                        break;
+
+                    default:
+                        parseCustomAffix(parser, builder);
+                        break;
+                }
+            } else if (tag == XmlPullParser.Event.END_ELEMENT) {
+                String name = parser.getName();
+                if (name.equals(ContentElement.ELEMENT) && ContentElement.NAMESPACE.equals(parser.getNamespace())) {
```

Comment:
I usually prefer the depth comparison, a single integer comparison, over two String equality checks.

Reply:
Via the QName?

Reply:
Ah, I see, checking the depth of the parser.